FFmpegInterop/dllmain.cpp

```diff
@@ -31,11 +31,24 @@ BOOL WINAPI DllMain(_In_ HINSTANCE hInstance, _In_ DWORD dwReason, _In_opt_ LPVO
 		// Register TraceLogging provider
 		TraceLoggingRegister(g_FFmpegInteropProvider);
 
-		// Register FFmpegInteropLogging callback
+#ifdef _DEBUG
+		// Register custom FFmpeg logging callback
 		av_log_set_callback(&FFmpegInteropLogging::Log);
+		av_log_set_level(AV_LOG_TRACE);
+#else
+		// Disable FFmpeg logging
+		av_log_set_callback(nullptr);
+		av_log_set_level(AV_LOG_QUIET);
+#endif // _DEBUG
 	}
 	else if (dwReason == DLL_PROCESS_DETACH)
 	{
+#ifdef _DEBUG
+		// Restore the default FFmpeg logging callback
+		// This is not thread safe! This could leave log_callback in av_vlog() dangling after this DLL is unloaded.
+		av_log_set_callback(av_log_default_callback);

```

Comment:
why isn't it thread safe? can't we fix ffmpeg?

Reply:
> why isn't it thread safe?

It's a classic readers-writers problem where the logging callback ([av_log_callback](https://github.com/search?q=repo%3AFFmpeg%2FFFmpeg+av_log_callback&type=code)) is the shared resource:
- Writer: thread setting the logging callback - [av_log_set_callback()](https://github.com/search?q=repo%3AFFmpeg%2FFFmpeg+symbol%3Aav_log_set_callback&type=code)
- Readers: threads invoking the logging callback - [av_vlog()](https://github.com/search?q=repo%3AFFmpeg%2FFFmpeg+symbol%3Aav_vlog&type=code)

You need to synchronize setting the logging callback with threads invoking the callback, otherwise the logging callback could become invalid (i.e. point to memory where our unloaded DLL used to be) while a thread is using it.

> can't we fix ffmpeg?

Yes, we could apply a patch to FFmpeg to address this problem by synchronizing av_log_set_callback() setting the logging callback with av_vlog() accessing the logging callback. This would allow the FFmpeg logging facility to be used safely with our custom logging callback.

Taking a quick peek at the FFmpeg code, it looks like they provide pthread_mutex_* equivalents using [SRWLOCK](https://learn.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks) in [w32pthreads.h](https://github.com/search?q=repo%3AFFmpeg%2FFFmpeg+path%3Aw32pthreads.h&type=code), but we would need to add pthread_rwlock_* equivalents using SRWLock.

Reply:
Note that adding synchronization around the logging callback would involve acquiring the writer lock in DllMain() when we call av_log_callback(). This can be perilous as it goes against [DLL best practices](https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-best-practices): "[You should never] synchronize with other threads [from within DllMain]. This can cause a deadlock."